### PR TITLE
fix #20972 Invalid and UB codegen for old-style case object transitions in refc

### LINKED ISSUE #20972: Invalid and UB codegen for old-style case object transitions
### Description

When using an old-style case object and assigning to its kind/discriminant, even when it's the same as the previous kind or with both discriminant branches identical or shared, Nim generates C code which writes past the end of the `kind` array it creates.

For example:
```nim
type ForkedEpochInfo = object
  case kind: bool
  of true, false: discard
var info = ForkedEpochInfo(kind: true)
doAssert info.kind
info.kind = false
doAssert not info.kind
```

Turns into (for example):
```c

N_LIB_PRIVATE N_NIMCALL(void, NimMainModule)(void) {
{
	NIM_BOOL T5_;
	nimZeroMem((void*)(&info__insufficient95space95for95an95object95of95type_3), sizeof(tyObject_ForkedEpochInfo__P9cjIwDL9ao00j59aOCgvNBBQ));
	info__insufficient95space95for95an95object95of95type_3.kind = NIM_TRUE;
	{
		if (!!(info__insufficient95space95for95an95object95of95type_3.kind)) goto LA3_;
		failedAssertImpl__stdZassertions_52(((NimStringDesc*) &TM__Y4lv6w4GBd353Yl9czI1zlw_2));
	}
	LA3_: ;
	T5_ = (NIM_BOOL)0;
	T5_ = NIM_FALSE;
(void*)NimDT___P9cjIwDL9ao00j59aOCgvNBBQ_kind);
	FieldDiscriminantCheck((NI)(NU)(info__insufficient95space95for95an95object95of95type_3.kind), (NI)(NU)(T5_), NimDT___P9cjIwDL9ao00j59aOCgvNBBQ_kind, 3);
	info__insufficient95space95for95an95object95of95type_3.kind = T5_;
	{
		if (!!(!(info__insufficient95space95for95an95object95of95type_3.kind))) goto LA8_;
		failedAssertImpl__stdZassertions_52(((NimStringDesc*) &TM__Y4lv6w4GBd353Yl9czI1zlw_3));
	}
	LA8_: ;
}
}
```

Where `info.kind = false` turns into that
```c
FieldDiscriminantCheck(
  (NI)(NU)(info__insufficient95space95for95an95object95of95type_3.kind),
  (NI)(NU)(T5_),
  NimDT___P9cjIwDL9ao00j59aOCgvNBBQ_kind,
  3);
```
call.

`FieldDiscrimantCheck` is
```c
N_LIB_PRIVATE N_NIMCALL(void, FieldDiscriminantCheck)(NI oldDiscVal, NI newDiscVal, TNimNode** a, NI L) {
	TNimNode* oldBranch;
	TNimNode* newBranch;
	oldBranch = selectBranch__system_6798(oldDiscVal, L, a);
	newBranch = selectBranch__system_6798(newDiscVal, L, a);
	{
		if (!!((newBranch == oldBranch))) goto LA3_;
		{
			if (!!((oldDiscVal == ((NI)0)))) goto LA7_;
			sysFatal__system_3084(((NimStringDesc*) &TM__Q5wkpxktOdTGvlSRo9bzt9aw_28));
		}
		goto LA5_;
		LA7_: ;
		{
			sysFatal__system_3084(((NimStringDesc*) &TM__Q5wkpxktOdTGvlSRo9bzt9aw_29));
		}
		LA5_: ;
	}
	LA3_: ;
}
```

Finally, `selectBranch__system_6798` is
```c
N_LIB_PRIVATE N_NIMCALL(TNimNode*, selectBranch__system_6798)(NI discVal, NI L, TNimNode** a) {
	TNimNode* result;
	result = (TNimNode*)0;
	result = a[(L)- 0];
	{
		NIM_BOOL T3_;
		TNimNode* x;
		T3_ = (NIM_BOOL)0;
		T3_ = ltpercent___system_1094(discVal, L);
		if (!T3_) goto LA4_;
		x = a[(discVal)- 0];
		{
			if (!!((x == ((TNimNode*) NIM_NIL)))) goto LA8_;
			result = x;
		}
		LA8_: ;
	}
	LA4_: ;
	return result;
}
```
This
```c
result = a[(L)- 0];
```
line is the final trigger of the issue.

The `NimDT___P9cjIwDL9ao00j59aOCgvNBBQ_kind` array is defined as:
```c
TNimNode* NimDT___P9cjIwDL9ao00j59aOCgvNBBQ_kind[3];
```

But then `FieldDiscriminantCheck` called with `L == 3`, which it passes along, with `a` to `selectBranch__system_6798`:
```c
	oldBranch = selectBranch__system_6798(oldDiscVal, L, a);
	newBranch = selectBranch__system_6798(newDiscVal, L, a);
```
so when `selectBranch__system_6798` runs, it tries to
```c
    result = a[(L)- 0];
```
but this is attempting to access the fourth element of  three-element array.

A final oddity is that ARC and ORC show neither the `CaseTransition` warning nor the UBSAN, output, but appear to function -- both `doAssert` work.

It's true that https://nim-lang.org/docs/manual.html#types-object-variants documents that this is deprecated, but (a) it seems inconsistent in this regard across GC types, and (b) it still should not generate invalid C code, regardless of anything else it does.

### Nim Version

```
Nim Compiler Version 1.7.3 [Linux: amd64]
Compiled at 2022-11-29
Copyright (c) 2006-2022 by Andreas Rumpf
```

```
Nim Compiler Version 1.6.10 [Linux: amd64]
Compiled at 2022-11-23
Copyright (c) 2006-2021 by Andreas Rumpf
```

```
Nim Compiler Version 1.6.9 [Linux: amd64]
Compiled at 2022-11-28
Copyright (c) 2006-2021 by Andreas Rumpf
```

### Current Output

```text
$ rm .cache/nim/ -rf && Nim/bin/nim c -r -d:release --mm:refc --passC="-fsanitize=undefined" --passL="-fsanitize=undefined" --passC=-Wall --passC=-Wextra --passC=-pedantic --passC=-flto --passL=-flto insufficient_space_for_an_object_of_type.nim 
insufficient_space_for_an_object_of_type.nim(6, 11) Warning: Potential object case transition, instantiate new object instead [CaseTransition]
.cache/nim/insufficient_space_for_an_object_of_type_r/@mNim@slib@ssystem.nim.c:4612:9: runtime error: load of address 0x5560a1c43af8 with insufficient space for an object of type '<unknown>'
0x5560a1c43af8: note: pointer points here
 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00
              ^ 
.cache/nim/insufficient_space_for_an_object_of_type_r/@mNim@slib@ssystem.nim.c:4612:9: runtime error: load of address 0x5560a1c43af8 with insufficient space for an object of type '<unknown>'
0x5560a1c43af8: note: pointer points here
 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00
              ^
```
```


### Expected Output

```text
Invalid C code not generated and no UBSAN output
```

### Possible Solution

_No response_

### Additional Information

Regarding accessing one element beyond the array bounds, but 6.5.6p8 addresses this:
> In other words, if the expression **P** points to the _i_-th element of an array object, the expressions **(P)+N** (equivalently, **N+(P)**) and **(P)-N** (where **N** has the value _n_) point to, respectively, the _i+n_-th and _i−n_-th elements of the array object, provided they exist. Moreover, if the expression **P** points to the last
element of an array object, the expression **(P)+1** points one past the last element of the array object, and if the expression **Q** points one past the last element of an array object, the expression **(Q)-1** points to the last element of the array object. If both the pointer operand and the result point to elements of the same array object, or one past the last element of the array object, the evaluation shall not produce an overflow; otherwise, the behavior is undefined. If the result points one past the last element of the array object, it shall not be used as the operand of a unary * operator that is evaluated.

So the existence of a pointer one element past the end of an array is well-defined and useful, but dereferencing it or accessing it is not.

### PR DESCRIPTION
**Pull request by bung87**
fix #20972

I was trying to figure out how to fix `FieldDiscriminantCheck`, in the end it's obvious no need to check as all case in one branch.

### CODE REVIEW

**Comment by Araq** [comment]
Too specific. And what do we do if the test starts to fail again? `refc` is on its way out.

**Comment by bung87** [comment]
so given a proper label to the issue ? seems "Won't Fix" not fit this situation.

**Comment by tersec** [comment]
As of https://github.com/nim-lang/Nim/commit/e278a781fc5bfc115326ed6c1873268d51b25303 even that specific example is still not fixed, regardless:
```
$ cat t20972.nim

discard """
  matrix: "--gc:refc; --gc:arc"
"""
{.passC: "-fsanitize=undefined -Wall -Wextra -pedantic -flto".}
{.passL: "-fsanitize=undefined -flto".}

type ForkedEpochInfo = object
  case kind: bool
  of true, false: discard
var info = ForkedEpochInfo(kind: true)
doAssert info.kind
info.kind = false
doAssert not info.kind
$ nim c -r -d:release --passC:-flto --passL:-flto --passC:-fsanitize=undefined --passL:-fsanitize=undefined --mm:refc t20972.nim
.cache/nim/t20972_r/@mNim@slib@ssystem.nim.c:4612:9: runtime error: load of address 0x5626ae301ef8 with insufficient space for an object of type '<unknown>'
0x5626ae301ef8: note: pointer points here
 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00
              ^
.cache/nim/t20972_r/@mNim@slib@ssystem.nim.c:4612:9: runtime error: load of address 0x5626ae301ef8 with insufficient space for an object of type '<unknown>'
0x5626ae301ef8: note: pointer points here
 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00
              ^
```
This PR just adds a test case, if looking for the actual UBSAN output, it'd need `-d:release` too (the codegen is invalid regardless, but UBSAN doesn't seem to pick up on it without `-d:release`)

**Comment by bung87** [comment]
now `cannot find -lubsan` error.

**Comment by tersec** [comment]
https://github.com/msys2/MINGW-packages/issues/3163
https://stackoverflow.com/questions/54027222/cannot-find-lubsan-on-using-fsanitize-undefined-mingw-w64

**Comment by arnetheduck** [comment]
@Araq keep in mind that `refc` needs to remain supported for bugfixes for at least 2 major releases from now to provide time for people to rewrite their code for the new GC, ie this is a significant amount of work imposed on the community and will need some time - in the meantime, we need to continue to treat refc as a first-class citizen and fix issues in it since the process will take months if not years.

**Comment by tersec** [comment]
https://github.com/nim-lang/Nim/pull/21103/commits/78c9867995bcd2bf2fd16958575f0d6fe1cac5db fixes at least this particular test case for me.

Now it's generating just assignments to the `kind` field, which seems fine:
```c
N_LIB_PRIVATE N_NIMCALL(void, NimMainModule)(void) {
{
        NIM_BOOL T5_;
        nimZeroMem((void*)(&info__t5048575550_3), sizeof(tyObject_ForkedEpochInfo__hqdJSUyEHspbutuqppJ5JQ));
        info__t5048575550_3.kind = NIM_TRUE;
        {
                if (!!(info__t5048575550_3.kind)) goto LA3_;
                failedAssertImpl__stdZassertions_52(((NimStringDesc*) &TM__LscqW5daIXoXqzgKM49a4rw_2));
        }
        LA3_: ;
        T5_ = (NIM_BOOL)0;
        T5_ = NIM_FALSE;
        info__t5048575550_3.kind = T5_;
        {
                if (!!(!(info__t5048575550_3.kind))) goto LA8_;
                failedAssertImpl__stdZassertions_52(((NimStringDesc*) &TM__LscqW5daIXoXqzgKM49a4rw_3));
        }
        LA8_: ;
}
}
```

File: compiler/semtypes.nim [hunk 1]
```diff
@@ -772,6 +772,8 @@ proc semRecordCase(c: PContext, n: PNode, check: var IntSet, pos: var int,
                  formatMissingEnums(c, a))
     else:
       localError(c.config, a.info, "not all cases are covered")
+  if n.len == 2 and chckCovered:
+    incl(a[0].sym.flags, sfDiscriminantAllInOne)
```
Comment (Araq) [inline compiler/semtypes.nim:776]: I don't understand this flag nor its implications.

Comment (bung87) [reply compiler/semtypes.nim:776]: It means users `of` branchs covered all case, no need inject `FieldDiscriminantCheck` code, might have better name though.

**Pull request closed without merging**